File: frontend/src/app/core/services/forum.service.ts
/**
 * Forum service.
 *
 * TODO list for junior developer:
 *   [ ] implement getPosts() – GET /forum/posts
 *   [ ] implement getPost() – GET /forum/posts/:id
 *   [ ] implement createPost() – POST /forum/posts
 *   [ ] implement reportPost() – POST /forum/posts/:id/report
 *   [ ] implement getInbox() – GET /messages
 *   [ ] implement sendMessage() – POST /messages
 *   [ ] implement getConversation() – GET /messages/:userId
 *   [ ] implement searchUsers() – GET /users/search?name=...
 */

import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';

import {
  ConversationSummary,
  DirectMessage,
  DirectMessageCreate,
  ForumPost,
  ForumPostCreate,
  ForumPostList,
  ReportCreate,
  UserPublic,
} from '../models';
import { ApiService } from './api.service';

@Injectable({ providedIn: 'root' })
export class ForumService {
  private readonly api = inject(ApiService);

  getPosts(page = 1, pageSize = 20): Observable<ForumPostList> {
    /**
     * TODO:
     *   return this.api.get<ForumPostList>(`/forum/posts?page=${page}&page_size=${pageSize}`);
     *
     * The backend automatically filters by the user's group+sector.
     * Do NOT add any client-side filtering.
     */
    throw new Error('getPosts() not yet implemented');
  }

  getPost(id: string): Observable<ForumPost> {
    /**
     * TODO:
     *   return this.api.get<ForumPost>(`/forum/posts/${id}`);
     */
    throw new Error('getPost() not yet implemented');
  }

  createPost(data: ForumPostCreate): Observable<ForumPost> {
    /**
     * TODO:
     *   return this.api.post<ForumPost>('/forum/posts', data);
     */
    throw new Error('createPost() not yet implemented');
  }

  reportPost(postId: string, data: ReportCreate): Observable<unknown> {
    /**
     * TODO:
     *   return this.api.post(`/forum/posts/${postId}/report`, data);
     */
    throw new Error('reportPost() not yet implemented');
  }

  // ──────────────────────────────────────────────────────────
  // Direct messages
  // ──────────────────────────────────────────────────────────

  getInbox(): Observable<ConversationSummary[]> {
    /**
     * TODO:
     *   return this.api.get<ConversationSummary[]>('/messages');
     */
    throw new Error('getInbox() not yet implemented');
  }

  sendMessage(data: DirectMessageCreate): Observable<DirectMessage> {
    /**
     * TODO:
     *   return this.api.post<DirectMessage>('/messages', data);
     */
    throw new Error('sendMessage() not yet implemented');
  }

  getConversation(userId: string, page = 1): Observable<DirectMessage[]> {
    /**
     * TODO:
     *   return this.api.get<DirectMessage[]>(`/messages/${userId}?page=${page}`);
     */
    throw new Error('getConversation() not yet implemented');
  }

  searchUsers(name: string): Observable<UserPublic[]> {
    /**
     * TODO:
     *   return this.api.get<UserPublic[]>(`/users/search?name=${encodeURIComponent(name)}`);
     *
     * Only users in the same group are returned (backend enforced).
     */
    throw new Error('searchUsers() not yet implemented');
  }
}

File: backend/app/services/forum_service.py
"""
Forum service.

⚠️  CRITICAL: Every query that returns content MUST apply the content filter.
    Never return posts that don't match the user's group+sector.
    The filter must be on the DB side – not in Python code after fetching all rows.

TODO list for junior developer:
  [ ] implement get_posts() – with content filter + pagination
  [ ] implement create_post()
  [ ] implement get_post_by_id() – verify user can see it
  [ ] implement search_users_for_dm() – name search within same group/sector
"""

from sqlalchemy import or_
from sqlalchemy.orm import Session

from app.core.constants import GroupVisibility, PostStatus, SectorVisibility
from app.models.forum import DirectMessage, ForumPost
from app.models.user import User
from app.schemas.forum import (
    DirectMessageCreate,
    ForumPostCreate,
    ForumPostListResponse,
    ForumPostResponse,
)


def _content_filter(query, current_user: User):
    """
    Apply the visibility filter to a query on ForumPost.

    A post is visible to the user if:
      (group_visibility == user.user_type OR group_visibility == "all")
      AND
      (sector_visibility == user.sector OR sector_visibility == "all")

    This filter is the heart of the privacy model – do not skip it!
    """
    # TODO: implement this helper and use it in every post query
    raise NotImplementedError("_content_filter() is not yet implemented")


def get_posts(
    db: Session,
    current_user: User,
    page: int = 1,
    page_size: int = 20,
) -> ForumPostListResponse:
    """
    Return a paginated list of posts visible to the current user.

    TODO:
      1. Start with db.query(ForumPost)
      2. Apply _content_filter(query, current_user)
      3. Filter status == VISIBLE
      4. Order by created_at DESC
      5. Apply offset + limit for pagination
      6. Return ForumPostListResponse
    """
    # TODO: implement this function
    raise NotImplementedError("get_posts() is not yet implemented")


def get_post_by_id(db: Session, post_id: str, current_user: User) -> ForumPost:
    """
    Return a single post – raise 404 if not found, 403 if not visible to user.

    TODO:
      1. Query ForumPost by id
      2. Check it passes _content_filter for this user
      3. Check status != DELETED
    """
    # TODO: implement this function
    raise NotImplementedError("get_post_by_id() is not yet implemented")


def create_post(db: Session, data: ForumPostCreate, author: User) -> ForumPost:
    """
    Create a new forum post.

    Validations:
      - Author must be ACTIVE
      - If group_visibility targets a specific group, it must match author's user_type
        (a widow cannot post in the widowers group)

    TODO:
      1. Validate visibility is not "broader" than author's actual group/sector
      2. Create ForumPost object
      3. db.add(), db.commit(), db.refresh()
      4. Return the post
    """
    # TODO: implement this function
    raise NotImplementedError("create_post() is not yet implemented")


def send_direct_message(
    db: Session, data: DirectMessageCreate, sender: User
) -> DirectMessage:
    """
    Send a private message.

    Validations:
      - Recipient must be in the same group as sender
      - Recipient must be ACTIVE
      - Sender is not blocked by recipient (check report history)

    TODO:
      1. Load recipient, validate same group
      2. Encrypt content (or mark as needing encryption)
      3. Create DirectMessage, save to DB
    """
    # TODO: implement this function
    raise NotImplementedError("send_direct_message() is not yet implemented")


def get_conversation(
    db: Session, current_user: User, other_user_id: str, page: int = 1
) -> list[DirectMessage]:
    """
    Return messages between current_user and other_user, newest first.

    TODO:
      1. Query DirectMessage where (sender=me AND recipient=other) OR (sender=other AND recipient=me)
      2. Order by created_at DESC
      3. Apply pagination (max 50 per page)
      4. Mark retrieved messages as is_read=True
    """
    # TODO: implement this function
    raise NotImplementedError("get_conversation() is not yet implemented")


def search_users_for_dm(db: Session, current_user: User, name: str) -> list[User]:
    """
    Search for users to send a DM to.

    Rules:
      - Only users in the SAME group as current_user
      - Search by first_name or last_name (case-insensitive)
      - Never expose contact details (phone/email) – name only

    TODO:
      1. Query users where user_type == current_user.user_type AND account_status == ACTIVE
      2. Filter by name ILIKE
      3. Return list (no PII beyond name)
    """
    # TODO: implement this function
    raise NotImplementedError("search_users_for_dm() is not yet implemented")
